File: scripts/changelog_sync.py
import os
import re
import subprocess
from datetime import datetime

ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
CHANGELOG = os.path.join(ROOT, "CHANGELOG.md")

def sh(cmd):
  return subprocess.check_output(cmd, cwd=ROOT, shell=True, text=True).strip()

def ensure_block(text):
  if "## [Unreleased]" not in text:
    text = text.strip() + "\n\n## [Unreleased]\n\n### Commits\n"
  if "### Commits" not in text:
    text = text.replace("## [Unreleased]", "## [Unreleased]\n\n### Commits")
  return text

def main():
  sha = sh("git rev-parse --short HEAD")
  msg = sh("git log -1 --pretty=%s")
  date = datetime.now().strftime("%Y-%m-%d")
  line = f"- {date} {msg} ({sha})"

  if os.path.exists(CHANGELOG):
    text = open(CHANGELOG, "r", encoding="utf-8").read()
  else:
    text = "# Changelog\n\n## [Unreleased]\n\n### Commits\n"

  text = ensure_block(text)
  if f"({sha})" in text:
    return 0

  m = re.search(r"## \[Unreleased\][\s\S]*?### Commits\s*\n", text)
  if not m:
    text = text.replace("## [Unreleased]", "## [Unreleased]\n\n### Commits\n")
    insert_at = text.find("### Commits") + len("### Commits\n")
    text = text[:insert_at] + line + "\n" + text[insert_at:]
  else:
    insert_at = m.end()
    text = text[:insert_at] + line + "\n" + text[insert_at:]

  open(CHANGELOG, "w", encoding="utf-8").write(text)
  print("CHANGELOG synced:", line)
  return 0

if __name__ == "__main__":
  raise SystemExit(main())

File: scripts/changelog-sync.mjs
/*
 * @Author: linyc
 * @Date: 2026-02-13 11:38:17
 * @LastEditTime: 2026-02-13 11:38:18
 * @LastEditors: linyc
 * @Description: 
 */
import fs from "fs";
import { execSync } from "child_process";

const file = "CHANGELOG.md";
const sha = execSync("git rev-parse --short HEAD").toString().trim();
const msg = execSync("git log -1 --pretty=%s").toString().trim();
const date = new Date().toISOString().slice(0, 10);
const line = `- ${date} ${msg} (${sha})`;

let content = fs.existsSync(file) ? fs.readFileSync(file, "utf8") : "# Changelog\n\n## [Unreleased]\n\n### Commits\n";
if (content.includes(`(${sha})`)) process.exit(0);

if (!content.includes("## [Unreleased]")) {
  content += "\n## [Unreleased]\n\n### Commits\n";
}
if (!content.includes("### Commits")) {
  content = content.replace("## [Unreleased]", "## [Unreleased]\n\n### Commits");
}

content = content.replace(/## \[Unreleased\]\n([\s\S]*?)(?=\n## \[|$)/, (m, body) => {
  if (!body.includes("### Commits")) body = `\n### Commits\n${body}`;
  return `## [Unreleased]\n${body}\n${line}\n`;
});

fs.writeFileSync(file, content, "utf8");
console.log("CHANGELOG synced:", line);